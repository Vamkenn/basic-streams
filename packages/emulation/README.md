# [@basic-streams](https://github.com/rpominov/basic-streams)/emulation

Although the version is above `1.0` this package is not stable. Please use with
caution API may change.
