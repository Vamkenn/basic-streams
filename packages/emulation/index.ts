import {Stream} from "@basic-streams/stream"

export class Value<T> {
  constructor(public readonly value: T) {}
}

export class TimeSpan {
  constructor(public readonly ms: number) {}
}

export type Timeline<T> = Array<TimeSpan | Value<T>>

export class Event<T> {
  constructor(
    public readonly time: number,
    public readonly value: T,
    public readonly cb?: ((x: T) => void),
  ) {}

  callCb() {
    if (this.cb) {
      this.cb(this.value)
    }
  }
}

export class EventsList<T> {
  constructor(public readonly items: Array<Event<T>>) {}

  static fromTimeline<T>(items: Timeline<T>, currentTime = 0): EventsList<T> {
    const events: Array<Event<T>> = []
    for (const item of items) {
      if (item instanceof TimeSpan) {
        currentTime = currentTime + item.ms
      } else {
        events.push(new Event(currentTime, item.value))
      }
    }
    return new EventsList(events)
  }

  withCb(cb: (value: T) => void): EventsList<T> {
    return new EventsList(
      this.items.map(item => new Event(item.time, item.value, cb)),
    )
  }

  merge(another: EventsList<T>): EventsList<T> {
    const events: Array<Event<T>> = []
    let indexA = 0
    let indexB = 0
    const itemsA = this.items
    const itemsB = another.items
    while (itemsA.length > indexA || itemsB.length > indexB) {
      if (
        itemsA.length !== indexA &&
        (itemsB.length === indexB || itemsA[indexA].time <= itemsB[indexB].time)
      ) {
        events.push(itemsA[indexA])
        indexA++
      } else {
        events.push(itemsB[indexB])
        indexB++
      }
    }
    return new EventsList(events)
  }

  takeOne(): {event: Event<T> | void; rest: EventsList<T>} {
    const [event, ...rest] = this.items
    return {event, rest: new EventsList(rest)}
  }

  toJSON() {
    return this.items.map(item => ({time: item.time, value: item.value}))
  }

  static jestSerializer = {
    test(x: any) {
      return x instanceof EventsList
    },

    // for outdated TS typings
    print(x: any): string {
      return ""
    },

    serialize<Refs, Config extends {indent: string}, T>(
      val: EventsList<T>,
      config: Config,
      indentation: string,
      depth: number,
      refs: Refs,
      printer: (
        x: T,
        config: Config,
        indentation: string,
        depth: number,
        refs: Refs,
      ) => string,
    ) {
      const separator = "\n"

      function printItem(item: Event<T>) {
        const indentation1 = indentation + config.indent
        const value = printer(item.value, config, indentation1, depth + 1, refs)
        return indentation1 + item.time + ": " + value
      }

      return `EventsList(${
        val.items.length === 0
          ? ""
          : separator + val.items.map(printItem).join(separator) + separator
      })`
    },
  }
}

export function t(ms: number): TimeSpan {
  return new TimeSpan(ms)
}

export function v<T>(x: T): Value<T> {
  return new Value(x)
}

export function emulate<T>(
  generator: (
    createStream: <U>(...timeline: Timeline<U>) => Stream<U>,
  ) => Stream<T>,
  maxTime = Infinity,
): EventsList<T> {
  const state = {
    time: 0,
    toProduce: new EventsList([]) as EventsList<any>,
    result: [] as Array<Event<T>>,
  }
  const resultStream = generator((...timeline) => {
    return cb => {
      // wrapping cb to make it a unique value
      // that we will compare to in unsuscribe
      const _cb = (x: any) => cb(x)
      state.toProduce = state.toProduce.merge(
        EventsList.fromTimeline(timeline, state.time).withCb(_cb),
      )
      return () => {
        state.toProduce = new EventsList(
          state.toProduce.items.filter(item => item.cb !== _cb),
        )
      }
    }
  })
  resultStream(value => {
    state.result.push(new Event(state.time, value))
  })
  while (true) {
    const {event, rest} = state.toProduce.takeOne()
    if (state.time >= maxTime || !event) {
      return new EventsList(state.result)
    }
    state.time = event.time
    state.toProduce = rest
    event.callCb()
  }
}

export function laterMock(
  createStream: <U>(...timeline: Timeline<U>) => Stream<U>,
) {
  return function later<T>(time: number, value?: T): Stream<T> {
    return createStream(t(time), v(value as any))
  }
}
