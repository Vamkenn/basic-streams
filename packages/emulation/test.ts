import {EventsList, emulate, t, v} from "./index"

expect.addSnapshotSerializer(EventsList.jestSerializer)

describe("emulate", () => {
  test("map", () => {
    expect(
      emulate(create => {
        return cb => create(t(10), v(1))(x => cb(x + 1))
      }),
    ).toMatchSnapshot()
  })

  test("filter", () => {
    expect(
      emulate(create => {
        return cb =>
          create(v(1), t(10), v(0), t(10), v(1))(x => {
            if (x > 0) cb(x)
          })
      }),
    ).toMatchSnapshot()
  })

  test("order of simultaneous values", () => {
    expect(
      emulate(create => {
        const a = create(t(10), v(1))
        const b = create(t(10), v(2))
        return cb => {
          const disposeA = a(cb)
          const disposeB = b(cb)
          return () => {
            disposeA()
            disposeB()
          }
        }
      }),
    ).toMatchSnapshot()
  })

  test("unsubscribe", () => {
    expect(
      emulate(create => {
        const a = create(t(10), v(1), t(10), v(0))
        const b = create(t(15), v(null))
        return cb => b(a(cb))
      }),
    ).toMatchSnapshot()
  })

  test("unsubscribe doesn't remove other values with same cb", () => {
    expect(
      emulate(create => {
        const a = create(t(10), v(1), t(10), v(0))
        const b = create(t(10), v(1), t(10), v(0))
        const unsubscriber = create(t(15), v(null))
        return cb => {
          a(cb)
          unsubscriber(b(cb))
          return () => {}
        }
      }),
    ).toMatchSnapshot()
  })
})
