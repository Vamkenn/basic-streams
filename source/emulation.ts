import {Stream} from "./index"

export class Value<T> {
  constructor(
    public readonly value: T,
    public readonly cb?: ((x: T) => void),
  ) {}

  callCb() {
    if (this.cb) {
      this.cb(this.value)
    }
  }

  toJSON() {
    return this.value
  }
}

export class TimeSpan {
  constructor(public readonly ms: number) {}

  toJSON() {
    return `after ${this.ms} ms...`
  }
}

type TimelineItem<T> = TimeSpan | Value<T>

export class Timeline<T> {
  constructor(
    public readonly items: ReadonlyArray<TimelineItem<T>>,
    dontCompact?: boolean,
  ) {
    if (dontCompact) {
      return this
    }

    const finalItems: Array<TimelineItem<T>> = []
    for (const item of this.items) {
      if (item instanceof TimeSpan) {
        if (item.ms === 0) {
          continue
        }
        if (finalItems.length > 0) {
          const last = finalItems[finalItems.length - 1]
          if (last instanceof TimeSpan) {
            finalItems[finalItems.length - 1] = new TimeSpan(last.ms + item.ms)
            continue
          }
        }
      }
      finalItems.push(item)
    }

    if (
      finalItems.length > 0 &&
      finalItems[finalItems.length - 1] instanceof TimeSpan
    ) {
      finalItems.pop()
    }

    return new Timeline(finalItems, true)
  }

  withCb(cb: (value: T) => void): Timeline<T> {
    return new Timeline(
      this.items.map(
        item => (item instanceof TimeSpan ? item : new Value(item.value, cb)),
      ),
    )
  }

  merge<U>(
    another: Timeline<U>,
    result: Timeline<T | U> = new Timeline([]),
  ): Timeline<T | U> {
    const a = this
    const b = another
    if (a.items.length === 0 && b.items.length === 0) {
      return result
    }
    if (a.items.length === 0) {
      return new Timeline([
        ...result.items,
        ...(b.items as Array<TimelineItem<T | U>>),
      ])
    }
    if (b.items.length === 0) {
      return new Timeline([
        ...result.items,
        ...(a.items as Array<TimelineItem<T | U>>),
      ])
    }
    const timeToValueA = a.timeToValue()
    const timeToValueB = b.timeToValue()
    if (timeToValueA === Infinity && timeToValueB === Infinity) {
      return new Timeline([
        new TimeSpan(Math.max(a.timeToEnd(), b.timeToEnd())),
        ...result.items,
      ])
    }
    const timeToValue = Math.min(timeToValueA, timeToValueB)
    const subtractA = a.subtractTime(timeToValue)
    const subtractB = b.subtractTime(timeToValue)
    return subtractA.rest.merge(
      subtractB.rest,
      new Timeline([
        ...result.items,
        new TimeSpan(timeToValue),
        ...(subtractA.values as Array<Value<T | U>>),
        ...(subtractB.values as Array<Value<T | U>>),
      ]),
    )
  }

  isEmpty(): boolean {
    return this.items.length === 0
  }

  takeOne(): {item: TimelineItem<T>; rest: Timeline<T>} {
    if (this.isEmpty()) {
      throw new Error("cannot takeOne() from an empty schedule")
    }
    const [top, ...rest] = this.items
    return {item: top, rest: new Timeline(rest)}
  }

  subtractTime(ms: number): {values: Value<T>[]; rest: Timeline<T>} {
    if (this.isEmpty()) {
      return {rest: this, values: []}
    }
    const {item, rest} = this.takeOne()
    if (item instanceof Value) {
      const r = rest.subtractTime(ms)
      return {rest: r.rest, values: [item, ...r.values]}
    }
    return item.ms <= ms
      ? rest.subtractTime(ms - item.ms)
      : {
          rest: new Timeline([new TimeSpan(item.ms - ms), ...rest.items]),
          values: [],
        }
  }

  timeToEnd(): number {
    return this.items.reduce(
      (result, x) => result + (x instanceof TimeSpan ? x.ms : 0),
      0,
    )
  }

  timeToValue(): number {
    if (this.isEmpty()) {
      return Infinity
    }
    const {item, rest} = this.takeOne()
    return item instanceof Value ? 0 : item.ms + rest.timeToValue()
  }

  toJSON() {
    return this.items
  }
}

type EmulationGenerator<T> = (
  createStream: <U>(...scheduleItems: Array<TimelineItem<U>>) => Stream<U>,
) => Stream<T>

export function t(ms: number): TimeSpan {
  return new TimeSpan(ms)
}

export function v<T>(x: T): Value<T> {
  return new Value(x)
}

export function emulate<T>(generator: EmulationGenerator<T>): Timeline<T> {
  const state = {
    schedule: new Timeline([]),
    result: [] as Array<TimelineItem<T>>,
  }

  const resultStream = generator((...scheduleItems) => {
    return cb => {
      state.schedule = state.schedule.merge(
        new Timeline(scheduleItems).withCb(cb),
      )
      return () => {
        state.schedule = new Timeline(
          state.schedule.items.filter(
            item => item instanceof TimeSpan || item.cb !== cb,
          ),
        )
      }
    }
  })

  resultStream(value => state.result.push(new Value(value)))

  while (state.schedule.timeToValue() !== Infinity) {
    const time = state.schedule.timeToValue()
    const {rest, values} = state.schedule.subtractTime(time)
    state.schedule = rest
    state.result.push(new TimeSpan(time))
    values.forEach(item => item.callCb())
  }

  return new Timeline(state.result)
}
