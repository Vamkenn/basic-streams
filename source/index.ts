export type Stream<T> = (cb: (payload: T) => void) => (() => void)
export type StreamLoose<T> = (
  cb: (payload: T, ...rest: any[]) => any,
  ...rest: any[]
) => any

type Maybe<T> = {tag: "nothing"} | {tag: "just"; value: T}

function noop() {}

function id<T>(x: T): T {
  return x
}

export function empty(): Stream<any> {
  return emptyStream
}
const emptyStream = () => noop

export function of<T>(x: T): Stream<T> {
  return cb => {
    cb(x)
    return noop
  }
}

export function fromIterable<T>(xs: Iterable<T>): Stream<T> {
  return cb => {
    for (const x of xs) {
      cb(x)
    }
    return noop
  }
}

export function fromLoose<T>(looseStream: StreamLoose<T>): Stream<T> {
  return _cb => {
    let cb: null | ((x: T) => any) = _cb
    let disposer = looseStream(x => {
      if (cb !== null) {
        cb(x)
      }
    })
    return () => {
      if (cb === null) {
        return
      }
      cb = null
      if (typeof disposer === "function") {
        disposer()
      }
      disposer = null
    }
  }
}

export function map<T, U>(fn: (x: T) => U, stream: Stream<T>): Stream<U> {
  return cb => stream(payload => cb(fn(payload)))
}

export function filter<T>(
  predicate: (x: T) => boolean,
  stream: Stream<T>,
): Stream<T> {
  return cb =>
    stream(payload => {
      if (predicate(payload)) cb(payload)
    })
}

export function chain<T, U>(
  fn: (x: T) => Stream<U>,
  stream: Stream<T>,
): Stream<U> {
  return cb => {
    let spawnedDisposers: Array<() => void> = []
    const mainDisposer = stream(payload => {
      spawnedDisposers.push(fn(payload)(cb))
    })
    return () => {
      spawnedDisposers.forEach(fn => fn())
      mainDisposer()
    }
  }
}

export function chainLatest<T, U>(
  fn: (x: T) => Stream<U>,
  stream: Stream<T>,
): Stream<U> {
  return cb => {
    let spawnedDisposer = () => {}
    const mainDisposer = stream(payload => {
      spawnedDisposer()
      spawnedDisposer = fn(payload)(cb)
    })
    return () => {
      spawnedDisposer()
      mainDisposer()
    }
  }
}

export function ap<T, U>(
  streamF: Stream<(x: T) => U>,
  streamV: Stream<T>,
): Stream<U> {
  return cb => {
    let latestF: Maybe<(x: T) => U> = {tag: "nothing"}
    let latestV: Maybe<T> = {tag: "nothing"}
    const push = () => {
      if (latestF.tag === "just" && latestV.tag === "just") {
        const fn = latestF.value
        cb(fn(latestV.value))
      }
    }
    const disposef = streamF(f => {
      latestF = {tag: "just", value: f}
      push()
    })
    const disposev = streamV(v => {
      latestV = {tag: "just", value: v}
      push()
    })
    return () => {
      disposef()
      disposev()
    }
  }
}

export function map2<A, B, C>(
  fn: (a: A, b: B) => C,
  streamA: Stream<A>,
  streamB: Stream<B>,
): Stream<C> {
  return ap(map((a: A) => (b: B) => fn(a, b), streamA), streamB)
}

export function map3<A, B, C, D>(
  fn: (a: A, b: B, c: C) => D,
  streamA: Stream<A>,
  streamB: Stream<B>,
  streamC: Stream<C>,
): Stream<D> {
  return ap(
    ap(map((a: A) => (b: B) => (c: C) => fn(a, b, c), streamA), streamB),
    streamC,
  )
}

export function combineArray<T>(streams: Array<Stream<T>>): Stream<Array<T>> {
  return streams.reduce(liftedAppend, of([]))
}
function liftedAppend<T>(xs: Stream<T[]>, x: Stream<T>): Stream<T[]> {
  return map2<T[], T, T[]>(append, xs, x)
}
function append<T>(xs: T[], x: T): T[] {
  return xs.concat([x])
}

export function merge<T>(streams: Array<Stream<T>>): Stream<T> {
  return chain<Stream<T>, T>(id, fromIterable(streams))
}

export function scan<N, A>(
  reducer: (acc: A, next: N) => A,
  seed: A,
  stream: Stream<N>,
): Stream<A> {
  return cb => {
    let acc = seed
    cb(acc)
    return stream(next => {
      acc = reducer(acc, next)
      cb(acc)
    })
  }
}

export function take<T>(n: number, stream: Stream<T>): Stream<T> {
  return cb => {
    let count = 0
    let disposer: (() => void) | null = null
    const dispose = () => {
      if (disposer !== null) {
        disposer()
        disposer = null
      }
    }
    disposer = stream(x => {
      count++
      if (count <= n) {
        cb(x)
      }
      if (count >= n) {
        dispose()
      }
    })
    if (count >= n) {
      dispose()
    }
    return dispose
  }
}

export function takeWhile<T>(
  predicate: (x: T) => boolean,
  stream: Stream<T>,
): Stream<T> {
  return cb => {
    let completed = false
    let disposer: (() => void) | null = null
    const dispose = () => {
      if (disposer !== null) {
        disposer()
        disposer = null
      }
    }
    disposer = stream(x => {
      if (!completed) {
        if (predicate(x)) {
          cb(x)
        } else {
          completed = true
          dispose()
        }
      }
    })
    if (completed) {
      dispose()
    }
    return dispose
  }
}

export function takeUntil<T>(
  controller: Stream<any>,
  stream: Stream<T>,
): Stream<T> {
  return cb => {
    let disposed = false
    let mainDisposer: (() => void) | null = null
    let ctrlDisposer: (() => void) | null = null
    const dispose = () => {
      if (disposed) {
        return
      }
      disposed = true
      if (mainDisposer !== null) {
        mainDisposer()
        mainDisposer = null
      }
      if (ctrlDisposer !== null) {
        ctrlDisposer()
        ctrlDisposer = null
      }
    }

    ctrlDisposer = controller(dispose)

    if (disposed && ctrlDisposer !== null) {
      ctrlDisposer()
    }

    if (disposed) {
      // subscribe anyway for consistency
      stream(noop)()
    } else {
      mainDisposer = stream(cb)
    }

    return dispose
  }
}

export function skip<T>(n: number, stream: Stream<T>): Stream<T> {
  return cb => {
    let count = 0
    return stream(x => {
      count++
      if (count > n) {
        cb(x)
      }
    })
  }
}

export function skipWhile<T>(
  predicate: (x: T) => boolean,
  stream: Stream<T>,
): Stream<T> {
  return cb => {
    let started = false
    return stream(x => {
      if (!started) {
        started = !predicate(x)
      }
      if (started) {
        cb(x)
      }
    })
  }
}

export function skipDuplicates<T>(
  comp: (prev: T, next: T) => boolean,
  stream: Stream<T>,
): Stream<T> {
  return sink => {
    let latest: Maybe<T> = {tag: "nothing"}
    return stream(x => {
      if (latest.tag === "nothing" || !comp(latest.value, x)) {
        latest = {tag: "just", value: x}
        sink(x)
      }
    })
  }
}

export function startWith<T, U>(x: T, stream: Stream<U>): Stream<T | U> {
  return cb => {
    cb(x)
    return stream(cb)
  }
}

export function multicast<T>(stream: Stream<T>): Stream<T> {
  let cbs: ((x: T) => void)[] = []
  const push = (x: T) => {
    cbs.forEach(sink => {
      if (cbs.indexOf(sink) !== -1) {
        sink(x)
      }
    })
  }
  let unsub: null | (() => void) = null
  return cb => {
    let disposed = false
    cbs = [...cbs, cb]
    if (cbs.length === 1) {
      unsub = stream(push)
    }
    return () => {
      if (disposed) {
        return
      }
      disposed = true
      const index = cbs.indexOf(cb)
      cbs = [...cbs.slice(0, index), ...cbs.slice(index + 1, cbs.length)]
      if (cbs.length === 0 && unsub !== null) {
        unsub()
        unsub = null
      }
    }
  }
}

/*

TODO:

later
interval
delay
throttle/debounce
sequentially (maybe add optional argument to fromIterable)

*/
