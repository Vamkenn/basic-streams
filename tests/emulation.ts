import {Value, TimeSpan, Timeline, emulate, t, v} from "../source/emulation"

describe("Timeline", () => {
  describe("merge", () => {
    test("empty", () => {
      const a = new Timeline([])
      const b = new Timeline([])
      expect(a.merge(b)).toMatchSnapshot()
    })

    test("two values", () => {
      const a = new Timeline([new Value(1)])
      const b = new Timeline([new Value(2)])
      expect(a.merge(b)).toMatchSnapshot()
    })

    test("two spans", () => {
      const a = new Timeline([new TimeSpan(10)])
      const b = new Timeline([new TimeSpan(5)])
      expect(a.merge(b)).toMatchSnapshot()
    })

    test("complex", () => {
      const a = new Timeline([new TimeSpan(10), new Value(3)])
      const b = new Timeline([
        new Value(1),
        new TimeSpan(5),
        new Value(2),
        new TimeSpan(5),
        new Value(4),
      ])
      expect(a.merge(b)).toMatchSnapshot()
    })
  })

  describe("constructor", () => {
    test("removes time tail", () => {
      expect(new Timeline([new Value(1), new TimeSpan(10)])).toMatchSnapshot()
    })

    test("removes 0s", () => {
      expect(
        new Timeline([new TimeSpan(0), new Value(1), new TimeSpan(0)]),
      ).toMatchSnapshot()
    })

    test("fuses time spans", () => {
      expect(
        new Timeline([new TimeSpan(5), new TimeSpan(10), new Value(1)]),
      ).toMatchSnapshot()
    })
  })
})

describe("emulate", () => {
  test("map", () => {
    expect(
      emulate(create => {
        return sink => create(t(10), v(1))(x => sink(x + 1))
      }),
    ).toMatchSnapshot()
  })

  test("filter", () => {
    expect(
      emulate(create => {
        return sink =>
          create(v(1), t(10), v(0), t(10), v(1))(x => {
            if (x > 0) sink(x)
          })
      }),
    ).toMatchSnapshot()
  })

  test("order of simultaneous values", () => {
    expect(
      emulate(create => {
        const a = create(t(10), v(1))
        const b = create(t(10), v(2))
        return sink => {
          const disposeA = a(sink)
          const disposeB = b(sink)
          return () => {
            disposeA()
            disposeB()
          }
        }
      }),
    ).toMatchSnapshot()
  })

  test("unsubscribe", () => {
    expect(
      emulate(create => {
        const a = create(t(10), v(1), t(10), v(0))
        const b = create(t(15), v(null))
        return sink => b(a(sink))
      }),
    ).toMatchSnapshot()
  })
})
