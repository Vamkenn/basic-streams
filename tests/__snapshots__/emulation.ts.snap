// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline constructor fuses time spans 1`] = `
Array [
  "after 15 ms...",
  1,
]
`;

exports[`Timeline constructor removes 0s 1`] = `
Array [
  1,
]
`;

exports[`Timeline constructor removes time tail 1`] = `
Array [
  1,
]
`;

exports[`Timeline merge complex 1`] = `
Array [
  1,
  "after 5 ms...",
  2,
  "after 5 ms...",
  3,
  4,
]
`;

exports[`Timeline merge empty 1`] = `Array []`;

exports[`Timeline merge two spans 1`] = `Array []`;

exports[`Timeline merge two values 1`] = `
Array [
  1,
  2,
]
`;

exports[`emulate filter 1`] = `
Array [
  1,
  "after 20 ms...",
  1,
]
`;

exports[`emulate map 1`] = `
Array [
  "after 10 ms...",
  2,
]
`;

exports[`emulate order of simultaneous values 1`] = `
Array [
  "after 10 ms...",
  1,
  2,
]
`;

exports[`emulate unsubscribe 1`] = `
Array [
  "after 10 ms...",
  1,
]
`;
